src/OrleansRuntime/GrainTypeManager/GrainTypeData.cs

```diff
@@ -151,5 +154,11 @@ internal static PlacementStrategy GetPlacementStrategy(Type grainClass, Placemen
 
             return predicate;
         }
+
+        private ParameterInfo[] GetConstructorParameters()
+        {
+            var constructor = this.Type.GetConstructors(BindingFlags.Public | BindingFlags.Instance).FirstOrDefault();
+            return constructor?.GetParameters() ?? EmptyParameters;
+        }
```

Comment:
>disallow custom activators or 3rd party containers that will choose the constructor in the way they want.

We don't support an application code only container, so same container will be used for all grains, Orleans, extensions, and application.  Constructor selection needs be a well defined behavior because it must be the same for internal Orleans grains, and grains used in extensions like the providers.  This is not a new constraint introduced by this PR.

>This is not compatible with supporting a different IGrainActivator

Yes it does.  Grain activators simply can't select constructors differently for non-application grains.  If we want to specifically make constructor selection part of the grain activator, we can do that, but that can only be supported for application grains, which is of little benefit compared to the nuance and complexity it would introduce.  It can also be added at later time if/when needed.

>A good approach is to do property injection

Early prototypes took this approach and were abandoned because constructor injection is more common, intuitive, testable, and fits into standard OO designs.

>You mentioned you were going to write up an issue with an extended design

I stated I'd write up an issue introducing the capability to the community.  I did not mention an extended design.  This system is still in development and will be improved over time as we refine it.